### PR TITLE
smoke: saturate a peer

### PR DESCRIPTION
**Pull request by FintanH**
Trying to replicate a dead-lock that we witness in a scenario between
a client and a seed node.

The real world scenario:
The client periodically creates a test every few seconds. On the 12th
project the seed does not replicate it. There's no output from logging
nor GIT_TRACE_PACKET.

This test scenario:
Peer 1 creates 12 projects and Peer 2 tries to replicate them all in a
loop. The dead-locking is not witnessed and the test passes.

Signed-off-by: Fintan Halpenny <fintan.halpenny@gmail.com>

### CODE REVIEW

**Comment by kim** [comment]
I think you forgot to add a file there..

File: librad/tests/smoke/saturation.rs
Comment (xla) [inline librad/tests/smoke/saturation.rs:17]: This needs to be 12+.

Comment (kim) [reply librad/tests/smoke/saturation.rs:17]: Is 12 a magic number there, yes?

Comment (FintanH) [reply librad/tests/smoke/saturation.rs:17]: There's the initial project too :) So it adds up to 12. Can bump it though if you think that'd make more sense?

**Comment by FintanH** [comment]
> I think you forgot to add a file there..

I did :grimacing: pushed now though :joy:

File: librad/tests/smoke/saturation.rs
Comment (FintanH) [reply librad/tests/smoke/saturation.rs:17]: > Is 12 a magic number there, yes?

Yup, the seed won't have any fetching logs nor git tracing when the 12th project was announced

Comment (kim) [reply librad/tests/smoke/saturation.rs:17]: My bets are on fd exhaustion...

**Comment by FintanH** [comment]
Going to try to emulate peer 1 gossiping the new projects and peer 2 replicating when they hear the gossip

File: librad/tests/smoke/saturation.rs
Comment (FintanH) [reply librad/tests/smoke/saturation.rs:17]: I believe @xla took a look at the fd's and didn't see anything out of the ordinary, but I'll let him chime in

File: librad/tests/smoke/saturation.rs
Comment (kim) [inline librad/tests/smoke/saturation.rs:80]: FYI: I removed this in my checkout. This branch is never created, so the gossip receiver will say “announced non-existent ref”, and refuse to forward the gossip message.

Comment (FintanH) [reply librad/tests/smoke/saturation.rs]: Ah :facepalm: thanks! I removed it here as well

File: librad/src/git/p2p/server.rs
Comment (kim) [inline librad/src/git/p2p/server.rs:219]: How about something like `io::copy(&mut recv, &mut io::sink())`?

File: librad/src/git/p2p/server.rs
Comment (alexjg) [inline librad/src/git/p2p/server.rs:229]: This won't happen if the `try_join` fails right? Should we save the `?` on the `try_join` result until after running `recv.read`?

**Review by alexjg — APPROVED** [review]

File: librad/src/git/p2p/server.rs
Comment (kim) [inline librad/src/git/p2p/server.rs:216]: Alright, trying to spell out what we want here:

* Drive `child` to completion
    
   There is `kill_on_drop`, but that's not a guarantee. That is, if `child` never terminates, the `server` never terminates. Applying some kind of backpressure on the creation of new git streams would not be a bad idea, but that's beyond the scope of this patch.

* Drive `copy(&mut stdout, &mut send)` to completion.

   Same thing: we need to drain `stdout` of the process.

* Drive `recv`, such that it receives either EOF or one byte of data (which, technically, would be a protocol violation).

* Drop all futures, or none at all


So, I think the `try_join` is good. The `copy(&mut recv, &mut devnull)` would continue to receive if the other end is a tarpit. If the other end is a tarpit, however, `receiving_unknown_size()` will be false, so dropping `recv` SHOULD be fine (in that, the stream gets deregistered).

This means that the previous attempt was better, except that the buffer should be 1 byte in length. If that 1 byte is not zero after `AsyncRead::poll_read`-ing once, this function should call the bad manners police immediately with the remote peer's id card.

I'm ok to merge this as-is for now, and follow up with a more elaborate thing.

**Comment by kim** [comment]
Only question here: do we really gain anything by running the test? I'm not sure it is the case that it would catch the underlying issue if it was re-introduced as a regression. As it's also not a cheap test, it means that it just creates heat.

If it's valuable to have it as a diagnostic/forensic tool, perhaps it could go into `e2e` as an executable?

**Comment by xla** [comment]
> Only question here: do we really gain anything by running the test? I'm not sure it is the case that it would catch the underlying issue if it was re-introduced as a regression. As it's also not a cheap test, it means that it just creates heat.

The smoke test does not complete since the number of projects was upped, which seems to force to hit the upper limit on the streams. I'd still consider it an accurate gauge if this problem is addressed or not, it doesn't hover show the relationship properly which is communicates poorly and will likely bamboozle new and current maintainers in six months alike.

**Review by kim — APPROVED** [review]

**Pull request merged**
